examples/find_optimal_communication_distance.py

```diff
@@ -34,72 +36,71 @@ def get_minimal_communication_distance(
 
-def build_placement_graph(
-    code: qldpc.codes.BBCode, folded_layout: bool, max_comm_dist: float
-) -> nx.Graph | None:
-    """Build a check qubit placement graph.  If some check qubit cannot be placed, return None.
+    # precompute the locations of all nodes' neighbors (which have fixed locations)
+    neighbor_locs = [
+        [code.get_qubit_pos(neighbor, folded_layout) for neighbor in graph.neighbors(node)]
+        for node in nodes
+    ]
 
-    The check qubit placement graph consists of two vertex sets:
-        (a) check qubits, and
-        (b) candidate locations.
-    The graph draws an edge betweeen qubit qq and location ll if qq's neighbors are at most
-    max_comm_dist away from ll.
-    """
-    nodes = [node for node in code.graph.nodes() if not node.is_data]
-    node_locs = [code.get_qubit_pos(node, folded_layout) for node in nodes]
+    # vectorized displacement calculation; shape = (len(nodes), len(locs), num_neighbors, 2)
+    displacements = (
+        np.array(locs, dtype=int)[None, :, None, :]
+        - np.array(neighbor_locs, dtype=int)[:, None, :, :]
+    )
 
-    def satisfies_max_comm_dist(node: qldpc.objects.Node, loc: tuple[int, int]) -> bool:
-        """Does placing a node at the given location satisfy the max_comm_dist constraint?"""
-        neighbors = set(code.graph.successors(node)).union(code.graph.predecessors(node))
-        return not any(
-            get_dist(loc, code.get_qubit_pos(neighbor, folded_layout)) > max_comm_dist
-            for neighbor in neighbors
-        )
+    distances_squared = np.sum(displacements**2, axis=-1)
+    placement_matrix = np.max(distances_squared, axis=-1)
 
-    graph = nx.Graph()
-    for node in nodes:
-        edges = [(node, loc) for loc in node_locs if satisfies_max_comm_dist(node, loc)]
-        if not edges:
-            return None
-        graph.add_edges_from(edges)
+    return nodes, locs, placement_matrix
 
-    return graph
 
+def get_perfect_matching(
+    vertices_a: Sequence[object],
+    vertices_b: Sequence[object],
+    biadjacency_matrix: npt.NDArray[np.bool_],
+) -> set[tuple[qldpc.objects.Node, tuple[int, int]]] | None:
+    """Find a perfect matching of a bipartite graph.  If no matching exists, return None."""
+    # quit early if any vertex has no indicent edges <--> any row/column is all zeros
+    if np.any(~np.any(biadjacency_matrix, axis=0)) or np.any(~np.any(biadjacency_matrix, axis=1)):
+        return None
+
+    # build the graph
+    graph = nx.Graph()
+    for idx_a, idx_b in zip(*np.where(biadjacency_matrix)):
+        graph.add_edge(vertices_a[idx_a], vertices_b[idx_b])
 
-@functools.cache
-def get_dist(loc_a: tuple[int, ...], loc_b: tuple[int, ...]) -> float:
-    """Euclidean (L2) distance between two locations."""
-    return math.sqrt(sum((aa - bb) ** 2 for aa, bb in zip(loc_a, loc_b)))
+    # find a perfect matching of the graph
+    matching = nx.bipartite.maximum_matching(graph, top_nodes=vertices_a)
```

Comment:
Passing a `top_nodes=...` argument to `nx.bipartite.maximum_matching` makes it work on disconnected graphs.